invoices/invoiceregistry.go

```diff
@@ -779,6 +779,56 @@ func (i *InvoiceRegistry) NotifyExitHopHtlc(rHash lntypes.Hash,
 	circuitKey channeldb.CircuitKey, hodlChan chan<- interface{},
 	payload Payload) (*HtlcResolution, error) {
 
+	// Execute locked notify exit hop logic.
+	resolution, err := i.notifyExitHopHtlcLocked(
+		rHash, amtPaid, expiry, currentHeight, circuitKey,
+		hodlChan, payload,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	switch r := resolution.(type) {
+
+	// A direct resolution was received for this htlc.
+	case *HtlcResolution:
+		return r, nil
+
+	// The htlc is held. Start a timer outside the lock if the htlc should
+	// be auto-released, because otherwise a deadlock may happen with the
+	// main event loop.
+	case *acceptResolution:
+		if r.autoRelease {
+			err := i.startHtlcTimer(rHash, circuitKey, r.acceptTime)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		return nil, nil
+
+	default:
+		return nil, errors.New("invalid resolution type")
+	}
+}
+
+// acceptResolution is returned when the htlc should be held.
+type acceptResolution struct {
+	// autoRelease signals that the htlc should be automatically released
+	// after a timeout.
+	autoRelease bool
+
+	// acceptTime is the time at which this htlc was accepted.
+	acceptTime time.Time
+}
+
+// notifyExitHopHtlcLocked is the internal implementation of NotifyExitHopHtlc
+// that should be executed inside the registry lock.
+func (i *InvoiceRegistry) notifyExitHopHtlcLocked(rHash lntypes.Hash,
+	amtPaid lnwire.MilliSatoshi, expiry uint32, currentHeight int32,
+	circuitKey channeldb.CircuitKey, hodlChan chan<- interface{},
+	payload Payload) (interface{}, error) {
+
```

Comment:
Can't comment below, but can't we restrict the scope of locking even further? All notifications could happen outside the lock right? That way this function extraction may not even be needed and the timer could stay also here.

Reply:
The notifications need to happen inside the lock to guarantee that they are delivered in order.

Reply:
Maybe I'm misunderstanding this but isn't delivery independent from locking? And therefore you cannot enforce delivery order with a lock?

Reply:
No, suppose we make update A and B in the database and because of no locking, the notifications arrive in the order B, A. Then it seems as if the invoice is in state A while it really is in B

Reply:
I understand that. To clarify, what I meant is that as notifications are sent by writing to channels, delivery order really is defined at the channel reader side. To answer my own question, if processing on the reader side is sequential, then the lock will indeed enforce order.

Reply:
Yes, reading is sequential indeed.